components/sensor/bme280/bme280.go

```diff
@@ -0,0 +1,579 @@
+		"dew_point_celsius":      dewPt,
+		"temperature_fahrenheit": temp*1.8 + 32,
+		"dew_point_fahrenheit":   dewPt*1.8 + 32,
+		"humidity_pct_rh":        humid,
+		"pressure_mpa":           pressure,
+	}, handle.Close()
+}
+
+// readPressure returns current pressure in mPa.
+func (s *bme280) readPressure(buffer []byte) float64 {
+	adc := float64((int(buffer[0])<<16 | int(buffer[1])<<8 | int(buffer[2])) >> 4)
+
+	// Need temp to calculate humidity
+	if s.lastTemp == -999 {
+		s.readTemperatureCelsius(buffer)
+	}
+
+	v1 := s.lastTemp/2. - 64000.
+	v2 := v1 * v1 * float64(s.calibration["digP6"]) / 32768.0
+	v2 += v1 * float64(s.calibration["digP5"]) * 2.
+	v2 = v2/4. + float64(s.calibration["digP4"])*65536.
+	v1 = (float64(s.calibration["digP3"])*v1*v1/524288.0 + float64(s.calibration["digP2"])*v1) / 524288.0
+	v1 = (1.0 + v1/32768.0) * float64(s.calibration["digP1"])
+
+	if v1 == 0 {
+		return 0
+	}
+
+	res := 1048576.0 - adc
+	res = ((res - v2/4096.0) * 6250.0) / v1
+	v1 = float64(s.calibration["digP9"]) * res * res / 2147483648.0
+	v2 = res * float64(s.calibration["digP8"]) / 32768.0
+	res += (v1 + v2 + float64(s.calibration["digP7"])) / 16.0
+	return res / 100.
+}
+
+// readTemperatureCelsius returns current temperature in celsius.
+func (s *bme280) readTemperatureCelsius(buffer []byte) float64 {
+	adc := float64((int(buffer[3])<<16 | int(buffer[4])<<8 | int(buffer[5])) >> 4)
+	var1 := (adc/16382 - float64(s.calibration["digT1"])/1024) * float64(s.calibration["digT2"])
+	var2 := math.Pow((adc/131072-float64(s.calibration["digT1"])/8192), 2) * float64(s.calibration["digT3"])
+
+	tFine := var1 + var2
+	s.lastTemp = tFine
+	output := tFine / 5120.
+	return output
+}
+
+// readHumidity returns current humidity as %RH.
+func (s *bme280) readHumidity(buffer []byte) float64 {
+	adc := float64(int(buffer[6])<<8 | int(buffer[7]))
+
+	// Need temp to calculate humidity
+	if s.lastTemp == -999 {
+		s.readTemperatureCelsius(buffer)
+	}
+
+	var1 := s.lastTemp - 76800.
+	var1 = (adc - (float64(s.calibration["digH4"])*64.0 + float64(s.calibration["digH5"])/16384.0*var1)) *
+		(float64(s.calibration["digH2"]) / 65536.0 * (1.0 + float64(s.calibration["digH6"])/67108864.0*var1*(1.0+
+			float64(s.calibration["digH3"])/67108864.0*var1)))
+	var1 *= (1.0 - (float64(s.calibration["digH1"]) * var1 / 524288.0))
+	return math.Max(0., math.Min(var1, 100.))
+}
+
+// calculateDewPoint returns current dew point in degrees C.
+func (s *bme280) calculateDewPoint(temp, humid float64) float64 {
+	ratio := 373.15 / (273.15 + temp)
+	rhs := -7.90298 * (ratio - 1)
+	rhs += 5.02808 * math.Log10(ratio)
+	rhs += -1.3816e-7 * (math.Pow(10, (11.344*(1-1/ratio))) - 1)
+	rhs += 8.1328e-3 * (math.Pow(10, (-3.49149*(ratio-1))) - 1)
+	rhs += math.Log10(1013.246)
+
+	// factor -3 is to adjust units - Vapor Pressure SVP * humidity
+	vp := math.Pow(10, rhs-3) * humid
+	// (2) DEWPOINT = F(Vapor Pressure)
+	t := math.Log(vp / 0.61078) // temp var
+	return (241.88 * t) / (17.558 - t)
```

Comment:
Not blocking but probably would be good to check NaN condition somewhere.

Reply:
In this particular case it's not mathematically possible for this to actually be NaN due to imitations on the byte widths of the inputs to temp and humidity. Nevertheless I will add a check.